### ISSUE TITLE
Cache load + compute methods are incompatible

### ISSUE THREAD

**Issue by ben-manes** [google/guava#2827]
A `Map.compute` may return `null` to indicate the entry should be removed if present. The `valueReference` is stored within the cache, allowing a subsequent `LoadingCache.get` to wait on it. This results in `LocalCache.waitForLoadingValue` to throw an `InvalidCacheLoadException` with the message `CacheLoader returned null for key`. Since the cache loader was not invoked, this message is incorrect and means the two methods cannot be used together, concurrently.

The user expectation would be that the load proceeds by retrying the load if `null` from a computation (but fails if from a cache loader). That may require using a sentinel value.

You can checkout this [commit](https://github.com/ben-manes/caffeine/commit/764df3aac6aacb78a4fc5a2134f911fd60dfa9f9) and run `MultiThreadedTest` using `./gradlew slowGuavaTest`. 

```java
Expected: is an empty collection
     but: <[Failed: key 15 on operation com.github.benmanes.caffeine.cache.MultiThreadedTest$$Lambda$3/841283083@91eee99
com.google.common.cache.CacheLoader$InvalidCacheLoadException: CacheLoader returned null for key 15.
	at com.google.common.cache.LocalCache$Segment.waitForLoadingValue(LocalCache.java:2321)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2206)
	at com.google.common.cache.LocalCache.get(LocalCache.java:4147)
	at com.google.common.cache.LocalCache.getAll(LocalCache.java:4210)
	at com.google.common.cache.LocalCache$LocalLoadingCache.getAll(LocalCache.java:5154)
	at com.github.benmanes.caffeine.cache.testing.GuavaCacheFromContext$GuavaLoadingCache.getAll(GuavaCacheFromContext.java:422)
	at com.github.benmanes.caffeine.cache.MultiThreadedTest.lambda$1(MultiThreadedTest.java:101)
	at com.github.benmanes.caffeine.testing.Threads$Thrasher.run(Threads.java:149)

**Comment by lowasser** [comment]
It sounds like this shouldn't be too difficult to deal with just by switching to a sentinel value, but if someone wants to send a pull request it'd probably get done more quickly.  Otherwise I'll try to take a whack at this...soon-ish.

**Comment by ben-manes** [comment]
Actually there are simpler fixes. A sentinel value would be invasive by requiring all read/write methods to check and translate, and computers to allocate a function to translate a null to the sentinel.

- Add a `ComputingValueReference` and have an instanceOf check in `waitForLoadingValue` when a null was detected. This class could extend `LoadingValueReference` as its merely tagging the behavioral difference.
- Add a decorator to the user's `CacheLoader` and move the assertion there. (This would also be needed in `get(key, callable)`).

All would require the cache loading calls to handle a retry loop. These two (or perhaps a similar variant) would be easy to implement and isolates the changes. We could also run a snapshot through the test suites to verify.

**Comment by JONGSHIN** [comment]
is this issue relevant?

**Comment by ben-manes** [comment]
@lowasser This is fixed in PR #5348 as part of fixing ##5342.
